node/core/dispute-coordinator/src/lib.rs

```diff
@@ -0,0 +1,649 @@
+			Err(e) => {
+				e.trace();
+
+				if let Error::Subsystem(SubsystemError::Context(_)) = e {
+					break;
+				}
+			}
+			Ok(()) => {
+				tracing::info!(target: LOG_TARGET, "received `Conclude` signal, exiting");
+				break;
+			}
+		}
+	}
+}
+
+// Run the subsystem until an error is encountered or a `conclude` signal is received.
+// Most errors are non-fatal and should lead to another call to this function.
+//
+// A return value of `Ok` indicates that an exit should be made, while non-fatal errors
+// lead to another call to this function.
+async fn run_iteration<Context>(ctx: &mut Context, subsystem: &DisputeCoordinatorSubsystem)
+	-> Result<(), Error>
+	where Context: SubsystemContext<Message = DisputeCoordinatorMessage>
+{
+	let DisputeCoordinatorSubsystem { ref store, ref keystore, ref config } = *subsystem;
+	let mut state = State {
+		keystore: keystore.clone(),
+		highest_session: None,
+		rolling_session_window: RollingSessionWindow::new(DISPUTE_WINDOW),
+	};
+
+	loop {
+		match ctx.recv().await? {
+			FromOverseer::Signal(OverseerSignal::Conclude) => {
+				return Ok(())
+			}
+			FromOverseer::Signal(OverseerSignal::ActiveLeaves(update)) => {
+				handle_new_activations(
+					ctx,
+					&**store,
+					&mut state,
+					config,
+					update.activated.into_iter().map(|a| a.hash),
+				).await?
+			}
+			FromOverseer::Signal(OverseerSignal::BlockFinalized(_, _)) => {},
+			FromOverseer::Communication { msg } => {
+				handle_incoming(
+					ctx,
+					&**store,
+					&mut state,
+					config,
+					msg,
+				).await?
+			}
+		}
+	}
+}
+
+async fn handle_new_activations(
+	ctx: &mut impl SubsystemContext,
+	store: &dyn KeyValueDB,
+	state: &mut State,
+	config: &Config,
+	new_activations: impl IntoIterator<Item = Hash>,
+) -> Result<(), Error> {
+	for new_leaf in new_activations {
+		let block_header = {
+			let (tx, rx) = oneshot::channel();
+
+			ctx.send_message(
+				ChainApiMessage::BlockHeader(new_leaf, tx).into()
+			).await;
+
+			match rx.await?? {
+				None => continue,
+				Some(header) => header,
+			}
+		};
+
+		match state.rolling_session_window.cache_session_info_for_head(
+			ctx,
+			new_leaf,
+			&block_header,
+		).await {
+			Err(e) => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					err = ?e,
+					"Failed to update session cache for disputes",
+				);
+
+				continue
+			}
+			Ok(SessionWindowUpdate::Initialized { window_end, .. })
+				| Ok(SessionWindowUpdate::Advanced { new_window_end: window_end, .. })
+			=> {
+				let session = window_end;
+				if state.highest_session.map_or(true, |s| s < session) {
+					tracing::trace!(
+						target: LOG_TARGET,
+						session,
+						"Observed new session. Pruning",
+					);
+
+					state.highest_session = Some(session);
+
+					db::v1::note_current_session(
+						store,
+						&config.column_config(),
+						session,
+					)?;
+				}
+			}
+			_ => {}
+		}
+
+		// TODO [after https://github.com/paritytech/polkadot/issues/3160]: chain rollbacks
```

Comment:
```suggestion
```

I guess this won't be happening anymore as we'll go with using `SelectChain` instead.